numpy/lib/_npyio_impl.py

```diff
@@ -252,8 +252,12 @@ def __getitem__(self, key):
             magic = bytes.read(len(format.MAGIC_PREFIX))
             bytes.close()
             if magic == format.MAGIC_PREFIX:
-                bytes = self.zip.open(key)
-                return format.read_array(bytes,
+                info = self.zip.NameToInfo[key]
+                assert info.compress_type == 0
+                self.zip.fp.seek(
+                    info.header_offset + len(info.FileHeader()) + 20
+                )
```

Comment:
A small nitpicks is that `NameToInfo` seems undocumented, so I think using `.getinfo()` is maybe nicer (unfortunately, so is `FileHeader()`).

One worry is that without duplicating `fp` first this isn't thread-safe or is `NpzFile` so fundamentally not thread-safe that we don't have to worry about it?

---

I am have wondering if we could clean up the `read_array` to use `fp.readinto(res_array)` to safe unnecessary copies.  Unfortunately, zipfile doesn't implement a specialized `readinto` so that only saves 1 of the two copies.  (which still may be nice on its own.)